File: iCyption2022/CftTest/Mati/solution.java
import java.util.*;

public class solution {

    static Scanner cin = new Scanner(System.in);

    public static int min(int a, int b) {
        if (a > b) 
            return b;
        return a;
    }

    static int INF = 1000000005;
    static ArrayList< ArrayList <Integer> > g;
    static ArrayList<Integer> timeFound;

    public static void dfs(int node, int prev, int timer) {
        timeFound.set(node, min(timeFound.get(node), timer));
        for (int child : g.get(node)) {
            if (child == prev) continue;
                dfs(child, node, timer + 1);
        }
    }

    public static void main(String[] args) {
        int n = cin.nextInt(), k = cin.nextInt();

        g = new ArrayList< ArrayList <Integer> > (n);
        timeFound = new ArrayList<Integer> (n);

        for (int i = 0; i < n; i++) timeFound.add(INF);
        for (int i = 0; i < n; i++) g.add(new ArrayList<Integer>());

        int x, y;
        for (int i = 0; i < n-1; i++) {
            x = cin.nextInt(); y = cin.nextInt(); x--; y--;
            g.get(x).add(y);
            g.get(y).add(x);
        }

        for (int i = 0; i < k; i++) {
            x = cin.nextInt(); x--;
            dfs(x, -1, 0);
        }

        int maxo = Collections.max(timeFound);
        ArrayList<Integer> ans = new ArrayList<Integer> ();

        for (int i = 0; i < n; i++)
            if (timeFound.get(i) == maxo) 
                ans.add(i + 1);

        System.out.println(ans.size());
        for (int a : ans)
            System.out.print(a + " ");
        }
}

File: iCyption2022/CftTest/Mati/solution.py
INF = 1000000005
g = []
timFound = []

def dfs(node, prev, timer) :
    timeFound[node] = min(timeFound[node], timer)
    for child in g[node]:
        if (child == prev) : continue
        dfs(child, node, timer + 1)

n, k = map(int, input().split())

g = [[] for i in range(n)]
timeFound = [INF for i in range(n)]

x, y = 0, 0
for i in range(n-1):
    x, y = map(int, input().split())
    x -= 1
    y -= 1
    g[x].append(y)
    g[y].append(x)

x = list(map(int, input().split()))
for i in x :
    i -= 1
    dfs(i, -1, 0)

maxo = max(timeFound)
ans = []

for i in range(n):
    if (timeFound[i] == maxo):
        ans.append(i + 1)

print(len(ans))
for x in ans:
    print(x, end=" ")
